landlab/utils/source_tracking_algorithm.py

```diff
@@ -0,0 +1,189 @@
+    """   
+    if flow_directions==None:
+        r = grid.at_node['flow__receiver_node']
+    else:
+        r = flow_directions
+    z = grid.at_node['topographic__elevation']
+    core_nodes = grid.core_nodes
+    core_elev = z[core_nodes]
+    # Sort all nodes in the descending order of elevation
+    sor_z = core_nodes[np.argsort(core_elev)[::-1]]
+    # Create a list to record all nodes that have been visited
+    # To store nodes that have already been counted
+    alr_counted = []
+    flow_accum = np.zeros(grid.number_of_nodes, dtype=int)
+    hsd_upstr = {}
+    # Loop through all nodes
+    for i in sor_z:
+        # Check 1: Check if this node has been visited earlier. If yes,
+        # then skip to next node
+        if i in alr_counted:
+            continue
+        # Check 2: If the visited node is a sink
+        if r[i] == i:
+            hsd_upstr.update({i: [hsd_ids[i]]})
+            flow_accum[i] += 1.
+            alr_counted.append(i)
+            continue
+        # Check 3: Now, if the node is not a sink and hasn't been visited, it
+        # belongs to a stream segment. Hence, all the nodes in the stream will
+        # have to betraversed.
+        # stream_buffer is a list that will hold the upstream contributing
+        # node information for that particular segment until reaching outlet.
+        stream_buffer = []
+        j = i
+        switch_i = True
+        a = 0.
+        # Loop below will traverse the segment of the stream until an outlet
+        # is reached.
+        while True:
+            # Following if loop is to execute the contents once the first node
+            # in the segment is visited.
+            if not switch_i:
+                j = r[j]
+                if j not in core_nodes:
+                    break
+            # If this node is being visited for the first time,
+            # this 'if statement' will executed.
+            if flow_accum[j] == 0.:
+                a += 1.
+                alr_counted.append(j)
+                stream_buffer.append(hsd_ids[j])
+            # Update number of upstream nodes.
+            flow_accum[j] += a
+            # If the node is being visited for the first time, the dictionary
+            # 'hsd_upstr' will be updated.
+            if j in hsd_upstr.keys():
+                hsd_upstr[j] += copy.copy(stream_buffer)
+            # If the node has been already visited, then the upstream segment
+            # that was not accounted for in the main stem, would be added to
+            # all downstream nodes, one by one, until the outlet is reached.
+            else:
+                hsd_upstr.update({j: copy.copy(stream_buffer)})
+            # If the outlet is reached, the 'while' loop will be exited.
+            if r[j] == j:
+                break
+            # This will be executed only for the first node of the
+            # stream segment.
+            if switch_i:
+                switch_i = False
+    return (hsd_upstr, flow_accum)
```

Comment:
Is this slow? It looks like it might need some optimization. Maybe not now but perhaps something to address in the future?